controllers/secretdefinition_controller.go

```diff
@@ -0,0 +1,242 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	secretsmanagerv1alpha1 "github.com/tuenti/secrets-manager/api/v1alpha1"
+	"github.com/tuenti/secrets-manager/backend"
+)
+
+const (
+	// https://golang.org/pkg/time/#pkg-constants
+	timestampFormat = "2006-01-02T15.04.05Z"
+	finalizerName   = "secret.finalizer.secrets-manager.tuenti.io"
+)
+
+// SecretDefinitionReconciler reconciles a SecretDefinition object
+type SecretDefinitionReconciler struct {
+	client.Client
+	Backend              backend.Client
+	Log                  logr.Logger
+	Ctx                  context.Context
+	APIReader            client.Reader
+	ReconciliationPeriod time.Duration
+}
+
+// Helper functions to check and remove string from a slice of strings.
+func containsString(slice []string, s string) bool {
+	for _, item := range slice {
+		if item == s {
+			return true
+		}
+	}
+	return false
+}
+
+func removeString(slice []string, s string) (result []string) {
+	for _, item := range slice {
+		if item == s {
+			continue
+		}
+		result = append(result, item)
+	}
+	return
+}
+
+// Ignore not found errors
+func ignoreNotFoundError(err error) error {
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	return err
+}
+
+// getDesiredState reads the content from the Datasource for later comparison
+func (r *SecretDefinitionReconciler) getDesiredState(keysMap map[string]secretsmanagerv1alpha1.DataSource) (map[string][]byte, error) {
+	desiredState := make(map[string][]byte)
+	var err error
+	for k, v := range keysMap {
+		bSecret, err := r.Backend.ReadSecret(v.Path, v.Key)
+		if err != nil {
+			r.Log.Error(err, "unable to read secret from backend", "path", v.Path, "key", v.Key)
+			return nil, err
+		}
+		decoder, err := backend.NewDecoder(v.Encoding)
+		if err != nil {
+			r.Log.Error(err, "refusing to use encoding", "encoding", v.Encoding)
+			return nil, err
+		}
+		desiredState[k], err = decoder.DecodeString(bSecret)
+		if err != nil {
+			r.Log.Error(err, "unable to decode data for secret", "encoding", v.Encoding, "path", v.Path, "key", v.Key)
+			return nil, err
+		}
+	}
+	return desiredState, err
+}
+
+// getCurrentState reads the content from the Kubernetes Secret API object for later comparison
+func (r *SecretDefinitionReconciler) getCurrentState(namespace string, name string) (map[string][]byte, error) {
+	// We don't read secrets from cache, as it's not the object we reconcile
+	reader := r.APIReader
+	data := make(map[string][]byte)
+	secret := &corev1.Secret{}
+	err := reader.Get(r.Ctx, client.ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}, secret)
+	if err != nil {
+		secretReadErrorsTotal.WithLabelValues(name, namespace).Inc()
+		return data, err
+	}
+	data = secret.Data
+	return data, err
+}
+
+// upsertSecret will create or update a secret
+func (r *SecretDefinitionReconciler) upsertSecret(secretDef *secretsmanagerv1alpha1.SecretDefinition, data map[string][]byte) error {
```

Comment:
xD extremely long line, github doens't show it completely

Reply:
alright. I'll try to make it shorter.